### PR TITLE
refactor(general): add epoch.Manager.MaybeCompactSingleEpoch

### PR DESCRIPTION
**Pull request by julio-lopez**
Subset and dependency of #3651
Depends on #3735

Ref:
- #3638
- #3639

### CODE REVIEW

File: internal/epoch/epoch_utils.go [hunk 1]
```diff
@@ -168,3 +168,49 @@ func getUncompactedEpochRange(cs CurrentSnapshot) (closedIntRange, error) {
...
+	return closedIntRange{
+		lo: cs.LongestRangeCheckpointSets[0].MinEpoch,
+		hi: cs.LongestRangeCheckpointSets[rangeSetsLen-1].MaxEpoch,
+	}
+}
+
+func oldestUncompactedEpoch(cs CurrentSnapshot) (int, error) {
+	rangeCompacted := getRangeCompactedRange(cs)
+
+	var oldesUncompactedEpoch int
```
Comment (plar) [inline internal/epoch/epoch_utils.go:186]: typo: oldestUncompactedEpoch

**Review by plar — APPROVED** [review]
LGTM

File: internal/epoch/epoch_manager.go [hunk 2]
```diff
@@ -970,6 +970,45 @@ func (e *Manager) getCompleteIndexSetForCommittedState(ctx context.Context, cs C
...
+
+	if !cs.isSettledEpochNumber(uncompacted) {
+		e.log.Debugw("there are no uncompacted epochs eligible for compaction", "oldestUncompactedEpoch", uncompacted)
+
+		return nil
+	}
+
+	uncompactedBlobs, ok := cs.UncompactedEpochSets[uncompacted]
+	if !ok {
+		// TODO: attempt to load blobs for this epoch
```
Comment (julio-lopez) [inline internal/epoch/epoch_manager.go:994]: argh, this is a stale left-over comment

File: internal/epoch/epoch_manager.go [hunk 3]
```diff
@@ -970,6 +970,45 @@ func (e *Manager) getCompleteIndexSetForCommittedState(ctx context.Context, cs C
 	return result, nil
 }
 
+// MaybeCompactSingleEpoch compacts the oldest epoch that is elegible for
```
Comment (redgoat650) [inline internal/epoch/epoch_manager.go:973]:
```suggestion
// MaybeCompactSingleEpoch compacts the oldest epoch that is eligible for
```

File: internal/epoch/epoch_manager_test.go [hunk 4]
```diff
@@ -800,6 +800,153 @@ func TestIndexEpochManager_RefreshContextCanceled(t *testing.T) {
...
+func TestMaybeCompactSingleEpoch_CompactionError(t *testing.T) {
+	t.Parallel()
+
+	te := newTestEnv(t)
+	ctx := testlogging.Context(t)
+
+	p, err := te.mgr.getParameters(ctx)
+	require.NoError(t, err)
+
+	// TODO: epochs, so there is an eligible epoch for compaction
```
Comment (redgoat650) [inline internal/epoch/epoch_manager_test.go:839]: Is this comment missing some words before "epochs,"? Seems like an incomplete thought

File: internal/epoch/epoch_manager.go [hunk 5]
```diff
@@ -970,6 +970,45 @@ func (e *Manager) getCompleteIndexSetForCommittedState(ctx context.Context, cs C
...
+	if err != nil {
+		return err
+	}
+
+	uncompacted, err := oldestUncompactedEpoch(cs)
+	if err != nil {
+		return err
+	}
+
+	if !cs.isSettledEpochNumber(uncompacted) {
```
Comment (redgoat650) [inline internal/epoch/epoch_manager.go:986]: So if I'm not mistaken, this has to do with the ordering of compaction vs epoch advancement we discussed, right? I.e. if we only advance epoch after successful compaction we only ever expect compaction-eligible epochs to be within the unsettled range.

**Review by redgoat650 — APPROVED** [review]
Some comment nits and a question, but looks good

File: internal/epoch/epoch_manager.go (hunk 5 above)
Comment (julio-lopez) [reply internal/epoch/epoch_manager.go:986]: @redgoat650 Yes, it's related. See the check right below.

In short, the current snapshot `cs` only loads the last 3 uncompacted epochs, that is, the current write epoch, and the previous 2. So, there may be some uncompacted epochs for which the blob metadata is not included in `cs`. The block below accounts for that and explicitly loads the list of the blobs for the "missing epochs".

File: internal/epoch/epoch_manager_test.go (hunk 4 above)
Comment (julio-lopez) [reply internal/epoch/epoch_manager_test.go:839]: It's a stale comment, it needs to be removed. I'll remove it in a minute.

**Pull request merged**